lib/test/Test/Pos/Ssc/Toss/BaseSpec.hs

```diff
@@ -594,28 +601,53 @@ checksBadCertsPayload (GoodPayload epoch sgs certsMap mrs) pk cert =
             _ -> qcFail $ "expected " <> show certNoRichmen <>
                           " to be a Left (CertificateNotRichmen ...)"
 
-        -- We take the VSS key of some cert from 'sgs' and replace a key of
-        -- some cert in 'certsMap' with it
-        res4 = fromMaybe (property True) $ do
-            c1 <- head (certs (sgs ^. sgsVssCertificates))
-            let c1id = addressHash . vcSigningKey $ c1
-            c2id <- head (HM.keys (getVssCertificatesMap certsMap))
-            let certsMap4 = certsMap
-                    & _Wrapped . ix c2id . _vcVssKey .~ vcVssKey c1
-                certDuplicateVss =
-                    tossRunner mrs sgs $
-                    checkCertificatesPayload epoch certsMap4
-            pure $ c1id /= c2id ==> case certDuplicateVss of
-                Left (CertificateDuplicateVssKey nes) -> c2id `qcElem` nes
-                _ -> qcFail $ "expected " <> show certDuplicateVss <>
-                              " to be a Left (CertificateDuplicateVssKey ...)"
-
-        allVssKeys = map vcVssKey (toList (getVssCertificatesMap certsMap))
+        getMapVssKeys :: VssCertificatesMap -> [AsBinary VssPublicKey]
+        getMapVssKeys = map vcVssKey . HM.elems . getVssCertificatesMap
+
+        certsMapVssKeys = getMapVssKeys certsMap
+
+        -- [CT-99]: VssCertificatesMap should not be allowed to contain a
+        -- certificate with a VSS key that already exists in VssCertData.
+        res4 = do let vssDataCertsMap = certs $ sgs ^. sgsVssCertificates
+                  let dataKeys  = getMapVssKeys vssDataCertsMap
+                  let anyInData = any (`elem` dataKeys) certsMapVssKeys -- is there key from map in VssData?
+                  not anyInData
 
     in (not (HM.member certSid $ mrs HM.! epoch) &&
-        not (vcVssKey cert `elem` allVssKeys))
+        not (vcVssKey cert `elem` certsMapVssKeys))
```

Comment:
I don't get this test now. It doesn't even call `checkCertificatesPayload`. But this test is about `checkCertificatesPayload`.
After reading previous version carefully, I think it was already good enough (it was testing the right thing, AFAIU).

Reply:
@gromakovsky Should I should rollback this change regarding `VssCertData` and keep only second test?

Reply:
Yes, I would just revert it.